peps/pep-0728.rst

```diff
@@ -0,0 +1,633 @@
+-------------------------
+
+It is an error to use ``Required[]`` or ``NotRequired[]`` with the special
+``__extra__`` field. ``total=False`` and ``total=True`` have no effect on
+``__extra__`` itself.
+
+The extra fields are non-required, regardless of the totality of the TypedDict.
+Operations that are available to ``NotRequired`` fields should also be available
+to the extra fields::
+
+    def f(movie: Movie) -> None:
+        del movie["name"]  # Not OK
+        del movie["year"]  # OK
+
+Interaction with ``Unpack``
+---------------------------
+
+For type checking purposes, ``Unpack[TypedDict]`` with extra fields should be
+treated as its equivalent in regular parameters, and the existing rules for
+function parameters still apply::
+
+    class Movie(TypedDict):
+        name: str
+        __extra__: int
+    
+    def f(**kwargs: Unpack[Movie]) -> None: ...
+
+    # Should be equivalent to
+    def f(*, name: str, **kwargs: int) -> None: ...
+
+Inheritance
+-----------
+
+``__extra__`` is inherited the same way as a regular ``key: value_type``
+field. The same rules from :pep:`PEP 589 <589#inheritance>` apply.
+
+Additionally, we need to reinterpret the following rule to define how
+``__extra__`` interacts with it:
+
+    * Changing a field type of a parent TypedDict class in a subclass is not allowed.
+
+First, it is not allowed to change the type of ``__extra__`` in a subclass::
+
+    class Parent(TypedDict):
+        __extra__: int | None
+    
+    class Child(Parent):
+        __extra__: int  # Not OK. Like any other TypedDict field, __extra__'s type cannot be changed
+
+Second, ``__extra__`` effectively defines the type of any unnamed fields
+accepted to the TypedDict and marks them as non-required. Thus, the above
+restriction applies to any additional fields defined in a subclass. For each
+field added in a subclass, all of the following conditions should apply:
+
+- The field is non-required
+
+- The field's value type is consistent with the type of ``__extra__``
+
+- The type of ``__extra__`` is consistent with the field's value type
+
+For example::
+
+    class MovieBase(TypedDict):
+        name: str
+        __extra__: int | None
+    
+    class AdaptedMovie(MovieBase):  # Not OK. 'bool' is not consistent with 'int | None'
+        adapted_from_novel: bool
+ 
+    class MovieRequiredYear(MovieBase):  # Not OK. Required key 'year' is not known to 'Parent'
+        year: int | None
+
+    class MovieNotRequiredYear(MovieBase):  # Not OK. 'int | None' is not consistent with 'int'
+        year: NotRequired[int]
+
+    class MovieWithYear(MovieBase):  # OK
+        year: NotRequired[int | None]
+
+Due to this nature, an important side effect allows us to define a TypedDict
+type that acts similarly to decorating the class definition with ``@final``::
+
+    class MovieFinal(TypedDict):
+        name: str
+        __extra__: Never
+
+Here, annotating ``__extra__`` with :class:`typing.Never` prevents ``MovieFinal``
```

Comment:
This isn't necessarily true under PEP 705 (since you could have a read-only item with a type you refine in a subclass). Also, pyright allows this (I think correctly):

```
from typing import Any, TypedDict

class TD1(TypedDict):
    x: list[Any]
    
class TD2(TD1):
    x: list[str]
```

More generally, I don't think "inheritance" is the right concept to talk about here. Because TypedDicts are structural types, inheritance isn't very important: you can get exactly the same type just by copy-pasting all the fields. What `__extra__: Never` does is say that there can be no other items in the TypedDict, whether in a subclass or otherwise.